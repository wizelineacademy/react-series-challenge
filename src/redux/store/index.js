import { createStore, applyMiddleware, compose } from 'redux';
import rootReducer from '../reducers';
import customMiddleware from '../middleware';
import createSagaMiddleware from 'redux-saga';
import rootSaga from '../sagas';

const sagaMiddleware = createSagaMiddleware();

const initialState = {};
const enhancers = [];
const middleware = [
  sagaMiddleware,
  ...customMiddleware,
];

const composedEnhancers = compose(
  applyMiddleware(...middleware),
  ...enhancers,
);

const store = createStore(
  rootReducer,
  initialState,
  composedEnhancers,
);

sagaMiddleware.run(rootSaga);

export default store;
