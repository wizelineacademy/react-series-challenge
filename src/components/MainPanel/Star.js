import React from 'react';

import starFav from './favorite.png';
import starNotFav from './addToFavorites.png';

import actions from '../../actions';
import { connect } from "react-redux";
import { bindActionCreators } from "redux";

import StyledStar from '../styled/Star.styled';

const { toggleFavorite } = actions.creators;

const Star = ({ active, id, toggleFavorite, gif }) => (
  <StyledStar 
    src={active? starFav:starNotFav}
    onClick={() => toggleFavorite(id, gif)}
  />
)

const mapDispatchToProps = dispatch => (
  bindActionCreators({ toggleFavorite }, dispatch)
)

export default connect(null, mapDispatchToProps)(Star);