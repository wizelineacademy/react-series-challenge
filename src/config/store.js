import {createStore, applyMiddleware, compose} from 'redux';
import createSagaMiddleware from 'redux-saga';

import rootReducer from '../reducers';
import rootSaga from '../sagas';
import {APPLICATION_LOCAL_STORAGE_KEY} from './config';
import {initialState as gifInitialState} from '../reducers/gif';

// Middlewares
const sagaMiddleware = createSagaMiddleware();
const middlewares = [
  sagaMiddleware
];

// Enhancers
const createEnhancers = () => {
  let enhancers = [];

  /**
   * Uncomment this lines and add the redux-devtools package to enable redux dev
   *  tools into development mode.
   */
  // if (process.env.NODE_ENV !== 'production') {
  //   enhancers.push(
  //     window.__REDUX_DEVTOOLS_EXTENSION__
  //       && window.__REDUX_DEVTOOLS_EXTENSION__());
  // }

  return enhancers;
};

const persistStateDetails = (state) => {
  try {
    let stateCopy = JSON.parse(JSON.stringify(state));

    // Remove unnecesary state details
    stateCopy.gifs.fetching = gifInitialState.fetching;
    stateCopy.gifs.displayed = gifInitialState.displayed;
    stateCopy.gifs.favoritesTextFilter = gifInitialState.favoritesTextFilter;

    let serializedState = JSON.stringify(stateCopy);
    localStorage.setItem(APPLICATION_LOCAL_STORAGE_KEY, serializedState);
  } catch (e) {}
}

const retrievePreviousStateDetails = () => {
  let result = {};

    try {
      let serializedState =
        window.localStorage.getItem(APPLICATION_LOCAL_STORAGE_KEY);

      if (serializedState) {
        result = JSON.parse(serializedState);
      }
    } catch (e) {}

  return result;
}

const enhancers = createEnhancers();
const composedEnhancers = (enhancers.length > 0)
  ? compose(applyMiddleware(...middlewares), ...enhancers)
  : applyMiddleware(...middlewares);


const initializeStore = (initialState) => {
  if (!initialState) {
    // Rehydrate the store
    initialState = retrievePreviousStateDetails();
  }

  // Store object
  const store = createStore(rootReducer, initialState, composedEnhancers);

  store.subscribe(() => persistStateDetails(store.getState()));

  sagaMiddleware.run(rootSaga);

  return store;
}

export default initializeStore;
