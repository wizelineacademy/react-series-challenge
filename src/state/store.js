import createSagaMiddleware from "redux-saga";
import { createStore, applyMiddleware } from "redux";

import rootSaga from "./sagas";
import reducers from "./reducers";

const sagaMiddleware = createSagaMiddleware();

const store = createStore(reducers, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(rootSaga);

export default store;
