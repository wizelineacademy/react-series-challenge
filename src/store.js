import { createStore, applyMiddleware, combineReducers } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import { gif, trending } from './reducers';

const reducer = combineReducers({ gif, trending });

export default createStore(reducer, composeWithDevTools(applyMiddleware(thunk)));
