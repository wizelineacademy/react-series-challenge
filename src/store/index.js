import { createStore, applyMiddleware, compose } from 'redux';
import rootReducer from '../reducers';
import initialState from '../reducers/initialState';
import createSagaMiddleware from "redux-saga";
import rootSaga from "../sagas";
import customMiddleware from '../middleware';

const sagaMiddleware = createSagaMiddleware();

const enhancers = [];
const middleware = [...customMiddleware, sagaMiddleware];

const composedEnhancers = compose(
  applyMiddleware(...middleware),
  ...enhancers,
);

const store = createStore(
  rootReducer,
  initialState,
  composedEnhancers,
);

sagaMiddleware.run(rootSaga);

export default store;
