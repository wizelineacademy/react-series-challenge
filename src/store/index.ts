import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga'
import { composeWithDevTools } from 'redux-devtools-extension';

import rootReducer from '../reducers';
import gifsFetcher from '../middleware';
import rootSaga from '../sagas/';
import autosave from '../middleware/autosave';

const sagaMiddleware = createSagaMiddleware();

function _getInitialFavs() {
  const data = localStorage.getItem('favs') || "";
  try {
    return JSON.parse(data);
  } catch(_) {
    return {}
  }
}

const initialState = {
  favs: _getInitialFavs()
};
const middleware = [
  sagaMiddleware,
  autosave,
  gifsFetcher,
];

const composedEnhancers = compose(
  applyMiddleware(...middleware),
);

const store = createStore(
  rootReducer,
  initialState,
  composeWithDevTools(composedEnhancers),
);

sagaMiddleware.run(rootSaga);

export default store;
