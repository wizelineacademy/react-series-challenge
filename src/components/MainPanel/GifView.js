import React from 'react';
import Star from './Star';

import StyledGifView, { Gif } from '../styled/GifView.styled.js';

const GifView = ({ src, id, favorite, gif }) => (
  <StyledGifView>
    <Gif src={src} />
    <Star active={favorite} id={id} gif={gif} />
  </StyledGifView>
)

export default GifView;