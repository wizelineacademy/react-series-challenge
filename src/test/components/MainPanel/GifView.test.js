import React from 'react';
import { mockGif } from "../../mocks";
import { shallow, mount } from "enzyme";
import { Provider } from "react-redux";

import GifView from "../../../components/MainPanel/GifView";
import StyledGifView, { Gif } from '../../../components/styled/GifView.styled';
import Star from '../../../components/MainPanel/Star';

import store from '../../../store';

describe('GifView', () => {
  
  it('GifView should have its children', () => {
    const comp = mount(
      <Provider store={store}>
        <GifView
          src={mockGif.url}
          id={mockGif.id}
          favorite={false}
          gif={mockGif}
        />
      </Provider>
    );
    expect(comp.find(StyledGifView)).toHaveLength(1);
    expect(comp.find(Gif)).toHaveLength(1);
    expect(comp.find(Star)).toHaveLength(1);
  })

  it('GifView should match snapshot', () => {
    const comp = shallow(
        <GifView
          src={mockGif.url}
          id={mockGif.id}
          favorite={false}
          gif={mockGif}
        />
    );
    expect(comp).toMatchSnapshot();
  })

})
