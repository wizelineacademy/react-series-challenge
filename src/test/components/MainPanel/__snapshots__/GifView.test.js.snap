// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifView GifView should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GifView
    favorite={false}
    gif={
        Object {
            "favorite": false,
            "id": "id",
            "original": "original",
            "title": "title",
            "url": "url",
          }
    }
    id="id"
    src="url"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <styled.img
          src="url"
/>,
        <Connect(Star)
          active={false}
          gif={
                    Object {
                              "favorite": false,
                              "id": "id",
                              "original": "original",
                              "title": "title",
                              "url": "url",
                            }
          }
          id="id"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "src": "url",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "active": false,
          "gif": Object {
            "favorite": false,
            "id": "id",
            "original": "original",
            "title": "title",
            "url": "url",
          },
          "id": "id",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <styled.img
            src="url"
/>,
          <Connect(Star)
            active={false}
            gif={
                        Object {
                                    "favorite": false,
                                    "id": "id",
                                    "original": "original",
                                    "title": "title",
                                    "url": "url",
                                  }
            }
            id="id"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "src": "url",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "gif": Object {
              "favorite": false,
              "id": "id",
              "original": "original",
              "title": "title",
              "url": "url",
            },
            "id": "id",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
